src/ted2zim/scraper.py

```diff
@@ -551,6 +629,9 @@ def update_videos_list(
                     "youtube_id": youtube_id,
                     "length": length,
                     "subtitles": subtitles,
+                    "subtitles_set": {
```

Comment:
Is there any reason to keep both the set and the list? I think what you've found is a bug and we should have only one property, and most probably transform "subtitles" to be a set (or a list of unique subtitles languages).

Reply:
Okay. Yes, I think you're correct. I had trouble understanding the logic for the function that generated subtitles list and looking at the output file, it sometimes contained duplicates. That was why I introduced a set to help keep it unique.
I will look further into it

Reply:
@benoit74 , I have examined the function that generates the subtitles list and  I think it is okay in the sense that it does what it's name implies and it also makes it suitable for testing purposes.
However, because it is being called repeatedly by other functions which make use of its results like `update_videos_list`, I think it is proper for users of such results to determine how to handle such results similar to how the `update_videos_list` only updates metadata if a video has been added before.
Because the members of the `subtitles` are `dict` entries, they are unhashable and thus, `subtitles` itself cannot be a set. Another way to deal with it would be to either use a `dataclass` or a custom type and implement the `__eq__` and `__hash__` logic but I think this is more complex and the above approach of maintaining an internal `subtitles_set` is simpler for the meantime.

Reply:
But where do we use this subtitles dict in your code? can't we use the set directly?

And a dict is a set with values, i.e. dict keys are know to be unique, just like set items.

Reply:
They're not used by my code but by the code that downloads the subtitles and encodes it with the video.

The logic for that is not implemented by me and I wanted to maintain the shape of the data.

Regarding uniqueness of dict values, the code that updates subtitles is `video[index]["subtitles"] += subtitles `
This is akin to a regular list `append` where each subtitle is a dict.

And filtering duplicates is not done in a list

Reply:
It is way more important to keep code clean and tidy, and avoid data duplication in memory (memory footprint is always an important thing). Of course it means you need to spend more time understanding code you're modifying, but this is the only long term solution to have a maintainable codebase and not a spaghetti.

That been said, `video[index]["subtitles"]` is a list, ok, but what is `subtitles` is it a dict`? If yes, then its keys are unique and you do not need the set (if dictionary key and set content are identical)

Reply:
No, `subtitles` is a list too returned from `generate_subtitles_list`.
I understand your point and I can see the error of my implementation as a way to circumvent a bug. The underlying issue is from the `generate_subtitles_list` function. 
There is this line:
```
        if self.subtitles_setting == ALL or (
            not self.source_languages and self.topics and self.subtitles_setting != NONE
        ):
            subtitles = [self.get_subtitle_dict(lang) for lang in langs]
```.
Now, because of the no language being set and a default `subtitle_setting` of `MATCHING`, this code will always generate all the available subtitles when run with the `--topics` flag because everything in the second part of the `OR` condition will be true.

Then, further in the `update_videos_list`, there is:

```
                if self.subtitles_setting in (MATCHING, NONE):
                    self.videos[index]["subtitles"] += subtitles
```
which I believe was meant to avoid duplicates but because of this new feature and the reason I explained above, it will always append everything to the `video[index]["subtitles"]` list as many times as each language was explored.

I will proceed to refactor the logic for the `generate_subtitles_list` function to account for these corner cases and avoid the use of the `subtitles_set`.

Reply:
I have removed the `subtitles_set` without refactoring the `generate_subtitles_list` as I believe that the function does what its name implies.

My fix for the issue was to refactor the condition that updated `subtitles` instead.
Prior to this feature, the logic was:

```
                if self.subtitles_setting in (MATCHING, NONE):
                    self.videos[index]["subtitles"] += subtitles
```
Now, I check the length explicitly instead:
```
                if self.subtitles_setting in (MATCHING, NONE) and len(subtitles) == 1:
                    self.videos[index]["subtitles"] += subtitles
```

Prior to this feature, the length of the subtitles list was always 1 in the conditions `MATCHING` and `NONE` as can be observed in this check in the body of the `generate_subtitles_list`.

```
        elif self.subtitles_setting == MATCHING or (
            self.subtitles_enough
            and self.subtitles_setting == NONE
            and page_lang != audio_lang
        ):
            subtitles = [
                self.get_subtitle_dict(lang)
                for lang in langs
                if lang["languageCode"] == page_lang
            ]
```

However, this feature caused the second part of the `OR` clause in the first condition to pass (detailed explanations are in the previous comment), 
```
        if self.subtitles_setting == ALL or (
            not self.source_languages and self.topics and self.subtitles_setting != NONE
        ):
            subtitles = [self.get_subtitle_dict(lang) for lang in langs]
```
thus, it generated the same `subtitles` list over and over again and introduced duplicates.

Now, I check it explicitly because even though the second part of the `OR` clause still causes the full subtitles list to be generated, we do not need to add it anymore as all will have already been generated.

All the other behaviours of the `subtitles_setting` still work as they should.

Hopefully, my explanations are clear.